modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/SparseVectorFieldMapper.java

```diff
@@ -0,0 +1,314 @@
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+
+    //*************STATIC HELPER METHODS******************************
+
+    // insert dim into the sorted array of dims
+    // insert value in the array of values in the same position as the corresponding dim
+    public static void sortedInsertDimValue(float[] values, int[] dims, float value, int dim, final int dimIndex) {
+        int newDimIndex = dimIndex;
+        while (newDimIndex > 0 && (dims[newDimIndex - 1] > dim)) {
+            // move higher valued dimension to the right
+            dims[newDimIndex] = dims[newDimIndex - 1] ;
+            values[newDimIndex] = values[newDimIndex - 1] ;
+            newDimIndex --;
+        }
+        dims[newDimIndex] = dim;
+        values[newDimIndex] = value;
+    }
+
+    /**
+     * Encodes a sparse array represented by values, dims and dimCount into a bytes array - BytesRef
+     * BytesRef: int[] floats encoded as integers values, 2 bytes for each dimension
+     * @param values - values of the sparse array
+     * @param dims - dims of the sparse array
+     * @param dimCount - number of the dimension
+     * @return BytesRef
+     */
+    private static BytesRef encodeSparseVector(float[] values, int[] dims, int dimCount) {
+        // 1. Sort dimensions in the ascending order and sort values in the same order as their corresponding dimensions
+        // as we expect that data should be already provided as sorted by dim,
+        // we expect just a single pass in this bubble sort algorithm
+        int temp;
+        float tempValue;
+        int n = dimCount;
+        boolean swapped = true;
+        while (swapped) {
+            swapped = false;
+            for (int i = 1; i <= n-1; i++){
+                if (dims[i-1] > dims[i]) {
+                    swapped = true;
+                    temp = dims[i];
+                    dims[i] = dims[i-1];
+                    dims[i-1] = temp;
+                    tempValue = values[i];
+                    values[i] = values[i-1];
+                    values[i-1] = tempValue;
+                }
+            }
+            n = n - 1;
+        }
+
+        byte[] buf = new byte[dimCount * (INT_BYTES + 2)];
+
+        // 2. Encode values
+        int offset = 0;
+        for (int dim = 0; dim < dimCount; dim++) {
+            int intValue = Float.floatToIntBits(values[dim]);
+            buf[offset] =  (byte) (intValue >> 24);
+            buf[offset+1] = (byte) (intValue >> 16);
+            buf[offset+2] = (byte) (intValue >>  8);
+            buf[offset+3] = (byte) intValue;
+            offset += INT_BYTES;
+        }
+
+        // 3. Encode dimensions
+        // as each dimension is a positive value that doesn't exceed 65000, 2 bytes is enough for encoding it
+        for (int dim = 0; dim < dimCount; dim++) {
+            buf[offset] = (byte) (dims[dim] >>  8);
+            buf[offset+1] = (byte) dims[dim];
+            offset += 2;
+        }
```

Comment:
not a big deal but it feels a bit weird to me to encode values before dimensions, it's a bit like encoding a value before its key?